File: 02Segundo/Programacion_Diseno_Orientada_Objetos_PDOO/Rupakalaki/lib/napakalaki/card_dealer.rb
module NapakalakiGame
  #  require 'pry'
  require 'singleton'
  require_relative 'treasure_kind'
  require_relative 'treasure'
  require_relative 'bad_consequence'
  require_relative 'numeric_bad_consequence'
  require_relative 'specific_bad_consequence'
  require_relative 'death_bad_consequence'
  require_relative 'cultist'
  require_relative 'prize'
  require_relative 'monster'

  class CardDealer
    include Singleton
    attr_accessor :usedMonsters, :unusedMonsters, :usedTreasures, :unusedTreasures, :usedCultists, :unusedCultists
    # 
    # Esta es una función que sirve para inicializar el mazo de los 
    # Tesoros se está intentando que se importen los tesoros
    # directamente desde un fichero de texto para que no tener que
    # crearlos todos a mano.
    #
    def initTreasureCardDeck()
      binding.pry if $DEBUGMODE
      @unusedTreasures = Array.new
      @usedTreasures = Array.new
      fd = File.open("#{$BD_TESOROS}", "r")
      fd.each_line do |line|
        columnas = line.split(",")
        @unusedTreasures.push(Treasure.new(columnas[0], columnas[4], columnas[2].to_i, columnas[3].to_i, columnas[1].upcase))
      end
      fd.close
      #      @unusedTreasures.each do |t| 
      #        puts "Tesoro: #{t.to_s}"
      #      end
      binding.pry if $DEBUGMODE
    end
    # 
    # Esta es una funcion que sirve para inicializar el mazo de los 
    # monstruos se esta intentando que se importen los monstruos
    # directamente desde un fichero de texto para que no tener que
    # crearlos todos a mano.
    #
    def initMonsterCardDeck()
      binding.pry if $DEBUGMODE
      @unusedMonsters = Array.new
      @usedMonsters = Array.new
      fd = File.open("#{$BD_MONSTRUOS}", "r")
      fd.each_line do |line|
        columnas = line.split(",")
        miPrice = Prize.new(columnas[2], columnas[3])
        if (columnas[10] == "true") then
          miBadConsequence = DeathBadConsequence.new(columnas[4], columnas[10])
        elsif (!columnas[8].empty? || !columnas[9].empty?) then
          #          tesorosVisibles = columnas[8].split("-")
          #          tesorosOcultos = columnas[9].split("-")
          tesorosVisibles = leeTesoros(columnas[8])
          tesorosOcultos = leeTesoros(columnas[9])
          miBadConsequence = SpecificBadConsequence.new(columnas[4], columnas[5].to_i, columnas[10], tesorosVisibles, tesorosOcultos) 
        else
          miBadConsequence = NumericBadConsequence.new(columnas[4], columnas[5].to_i, columnas[10], columnas[6].to_i, columnas[7].to_i)
        end
        @unusedMonsters.push(Monster.new(columnas[0], columnas[1].to_i, miBadConsequence, miPrice))
      end
      fd.close
      #      @unusedMonsters.each do |m| 
      #        puts "Monstruo: #{m.to_s}"
      #      end
      binding.pry if $DEBUGMODE
    end
    # 
    # Esta es una funcion que sirve para inicializar el mazo de los 
    # Jugadores Sectarios se importen directamente desde un fichero
    # de texto para que no tener que crearlos todos a mano.
    #
    def initCultistCardDeck()
      binding.pry if $DEBUGMODE
      @unusedCultists = Array.new
      @usedCultists = Array.new
      fd = File.open("#{$BD_SECTARIOS}", "r")
      fd.each_line do |line|
        columnas = line.split(",")
        @unusedCultists.push(Cultist.new(columnas[0], columnas[1].to_i))
      end
      fd.close
      binding.pry if $DEBUGMODE
    end
    
    def shuffleTreasures()
      @unusedTreasures = @unusedTreasures.shuffle
    end
  
    def shuffleMonsters()
      @unusedMonsters = @unusedMonsters.shuffle
    end
  
    def shuffleCultists()
      @unusedCultists = @unusedCultists.shuffle
    end
    
    def nextTreasure()
      binding.pry if $DEBUGMODE
      if @unusedTreasures.empty?
        @usedTreasures.each do |t| 
          @unusedTreasures<<t
        end
        shuffleTreasures()
        @usedTreasures.clear
      end
      t = @unusedTreasures.at(0)
      @usedTreasures<<t
      @unusedTreasures.delete(t);
      binding.pry if $DEBUGMODE
      return t
    end
  
    def nextMonster()
      binding.pry if $DEBUGMODE
      if @unusedMonsters.empty?
        @usedMonsters.each do |m| 
          @unusedMonsters<<m
        end
        shuffleMonsters()
        @usedMonsters.clear
      end
      m = @unusedMonsters.at(0)
      @usedMonsters<<m
      @unusedMonsters.delete(m);
      binding.pry if $DEBUGMODE
      return m
    end
  
    def nextCultist()
      binding.pry if $DEBUGMODE
      if @unusedCultists.empty?
        @usedCultists.each do |m| 
          @unusedCultists<<m
        end
        shuffleCultists()
        @usedCultists.clear
      end
      c = @unusedCultists.at(0)
      @usedCultists<<c
      @unusedCultists.delete(c);
      binding.pry if $DEBUGMODE
      return c
    end
    
    def giveTreasuresBack(t)
      @usedTreasures<<t
    end
  
    def giveMonstersBack(m)
      @usedMonsters<<m
    end
    
    def giveCardBack(c)
      @usedMonsters<<c
    end
  
    def initCards()
      binding.pry if $DEBUGMODE
      initTreasureCardDeck()
      shuffleTreasures()
      initMonsterCardDeck()
      shuffleMonsters()
      initCultistCardDeck()
      shuffleCultists()
      binding.pry if $DEBUGMODE
    end
  
    def leeTesoros(cadena_tesoros)
      cadena_tesoros = cadena_tesoros.upcase
      cadena_tesoros = cadena_tesoros.split("-")
      binding.pry if $DEBUGMODE
      # estos serán del tipo Treasure Kind
      tesoros = Array.new
      cadena_tesoros.each do |tk|
        case tk
        when "ARMOR" then
          tesoros << TreasureKind::ARMOR
        when "ONEHAND" then
          tesoros << TreasureKind::ONEHAND
        when "BOTHHANDS" then
          tesoros << TreasureKind::BOTHHANDS
        when "HELMET" then
          tesoros << TreasureKind::HELMET
        when "SHOES" then
          tesoros << TreasureKind::SHOES
        end
      end
      binding.pry if $DEBUGMODE
      return tesoros
    end
    
    private :initMonsterCardDeck, :initTreasureCardDeck, :initCultistCardDeck, :shuffleMonsters, :shuffleTreasures, :shuffleCultists, :leeTesoros
  end
end

File: 02Segundo/Programacion_Diseno_Orientada_Objetos_PDOO/Napakalaki/src/napakalaki/CardDealer.java
package napakalaki;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;

/**
 * Clase principal del juego, contiene las instancias principales para el
 * desarrollo de una partida.
 *
 * @author: Carlos de la Torre
 */
public class CardDealer {

    /**
     * Instance es la unica instancia del mazo de cartas que hay en toda la
     * aplicacion de esta manera no se pueden repetir mazos de cartas, esto
     * sirve para las clases singleton.
     */
    private static final CardDealer INSTANCE = new CardDealer();
    /**
     * Estas variables son para el gestor de ficheros, se usan para poder leer
     * los ficheros de textos necesarios para cargar todos los monstruos,
     * tesoros y demas cartas en los mazos correspondientes.
     */
    private FileReader fr = null;
    /**
     * Estas variables son para el gestor de ficheros, se usan para poder leer
     * los ficheros de textos necesarios para cargar todos los monstruos,
     * tesoros y demas cartas en los mazos correspondientes.
     */
    private BufferedReader br = null;
    /**
     * Estas variables se usan para cuando se realizan las lecturas de los
     * ficheros de texto se puedan guardar en un modelo de datos tipo matriz.
     */
    private String[] columnas = null;
    /**
     * Estas variables se usan para cuando se realizan las lecturas de los
     * ficheros de texto se puedan guardar en un modelo de datos tipo matriz.
     */
    private String fila = null;
    /**
     * Este es el mazo de cartas que corresponde a los monstruos que NO se están
     * usando en la partida.
     */
    public ArrayList<Monster> mUnusedMonsters;
    /**
     * Este es el mazo de cartas que corresponde a los monstruos que se están
     * usando en la partida.
     */
    public ArrayList<Monster> mUsedMonsters;
    /**
     * Este es el mazo de cartas que corresponde a los tesoros que NO se están
     * usando en la partida.
     */
    public ArrayList<Treasure> mUnusedTreasures;
    /**
     * Este es el mazo de cartas que corresponde a los tesoros que se están
     * usando en la partida.
     */
    public ArrayList<Treasure> mUsedTreasures;
    /**
     * Este es el mazo de cartas que corresponde a los sectarios que NO se están
     * usando en la partida.
     */
    public ArrayList<Cultist> mUnusedCultists;
    /**
     * Este es el mazo de cartas que corresponde a los sectarios que se están
     * usando en la partida.
     */
    public ArrayList<Cultist> mUsedCultists;

    /**
     * Es el constructor por defecto, pero está privado por que vamos a usar la
     * clase de manera que sea una clase singleton, osea que solo habrá una
     * instancia de esta clase en todo el juego.
     */
    private CardDealer() {
        this.mUnusedMonsters = new ArrayList();
        this.mUsedMonsters = new ArrayList();
        this.mUsedTreasures = new ArrayList();
        this.mUnusedTreasures = new ArrayList();
        this.mUsedCultists = new ArrayList();
        this.mUnusedCultists = new ArrayList();
    }

    /**
     * Función auxiliar que se usa para poder leer el fichero de configuración
     * de los diferentes mazos de cartas cuando se usa el juego en un fichero
     * compilado .jar.
     *
     * @param recurso esta es una cadena de texto que será donde se encuentra el
     * fichero
     * @return devuelve un objeto File con el fichero que le hemos pasado por
     * parametros
     */
    private File ficheroExterno(String recurso) {
        File fichero = null;
        URL res = getClass().getResource(recurso);
        if (res.toString().startsWith("jar:")) {
            try {
                InputStream input = getClass().getResourceAsStream(recurso);
                fichero = File.createTempFile("tempfile", ".tmp");
                OutputStream out = new FileOutputStream(fichero);
                int read;
                byte[] bytes = new byte[1024];

                while ((read = input.read(bytes)) != -1) {
                    out.write(bytes, 0, read);
                }
                fichero.deleteOnExit();
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        } else {
            //this will probably work in your IDE, but not from a JAR
            fichero = new File(res.getFile());
        }

        if (fichero != null && !fichero.exists()) {
            throw new RuntimeException("Error: File " + fichero + " not found!");
        }
        return fichero;
    }

    /**
     * Inicializa el monton de cartas de Tesoros para que se puedan usar en el
     * juego las carga desde un fichero de texto y las baraja.
     */
    private void initTreasureCardDeck() {
        try {
            // Apertura del fichero y creacion de BufferedReader para poder
            // hacer una lectura comoda (disponer del metodo readLine()).
            //String fichero = getClass().getResource("/resources/base_datos_tesoros.txt").getPath(); // < para Linux desde el IDE
            //String fichero = "C:\\Users\\pc\\Desktop\\Napakalaki\\src\\resources\\base_datos_tesoros.txt"; // < para Windows
            String fichero = "/resources/base_datos_tesoros.txt"; // < para Linux desde fichero JAR
            File file = this.ficheroExterno(fichero); // Creamos el fichero
            fr = new FileReader(file); // Creamos el manejador de ficheros
            br = new BufferedReader(fr); // Creamos el buffer de lectura para el fichero
            br.readLine(); // lo usamos para leer las cabeceras de las columnas 
            while ((fila = br.readLine()) != null) {
                columnas = fila.split(",");
                mUnusedTreasures.add(new Treasure(columnas[0], Integer.parseInt((columnas[4])), Integer.parseInt(columnas[2]), Integer.parseInt(columnas[3]), TreasureKind.valueOf(columnas[1].toUpperCase())));
            }
            this.shuffleTreasures();
            fr.close();
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
    }

    /**
     * Inicializa el monton de cartas de Monstruos para que se puedan usar en el
     * juego las carga desde un fichero de texto y las baraja.
     */
    private void initMonsterCardDeck() {
        try {
            BadConsequence malrollo = null;
            // Apertura del fichero y creacion de BufferedReader para poder
            // hacer una lectura comoda (disponer del metodo readLine()).
            //String fichero = getClass().getResource("/resources/base_datos_monstruos.txt").getPath(); // < para Linux desde el IDE
            //String fichero = "C:\\Users\\pc\\Desktop\\Napakalaki\\src\\resources\\base_datos_monstruos.txt"; // < para Windows
            String fichero = "/resources/base_datos_monstruos.txt"; // < para Linux desde fichero JAR
            File file = this.ficheroExterno(fichero); // Creamos el fichero
            fr = new FileReader(file); // Creamos el manejador de ficheros
            br = new BufferedReader(fr); // Creamos el buffer de lectura para el fichero
            br.readLine(); // lo usamos para leer las cabeceras de las columnas 
            while ((fila = br.readLine()) != null) {
                columnas = fila.split(",");
                if (columnas[10].equals("true")) {
                    malrollo = new DeathBadConsequence(columnas[4], true);
                } else if (!columnas[8].isEmpty() || !columnas[9].isEmpty()) {
                    ArrayList<TreasureKind> THidden = this.leeTesoros(columnas[8]);
                    ArrayList<TreasureKind> TVisible = this.leeTesoros(columnas[9]);
                    malrollo = new SpecificBadConsequence(columnas[4], Integer.parseInt(columnas[5]), false, THidden, TVisible);
                } else {
                    malrollo = new NumericBadConsequence(columnas[4], Integer.parseInt(columnas[5]), false, Integer.parseInt(columnas[6]), Integer.parseInt(columnas[7]));
                }
                Prize precio = new Prize(Integer.parseInt(columnas[2]), Integer.parseInt(columnas[3]));
                mUnusedMonsters.add(new Monster(columnas[0], Integer.parseInt(columnas[1]), malrollo, precio, Integer.parseInt(columnas[11])));
            }
            this.shuffleMonsters();
            fr.close();
        } catch (IOException | NumberFormatException e) {
            e.printStackTrace();
            System.exit(0);
        }
    }

    /**
     * Con este metodo lo que hacemos es cargar en el programa todas las cartas
     * necesarias para convertir a un jugador en sectario.
     */
    private void initCultistCardDeck() {
        try {
            // Apertura del fichero y creacion de BufferedReader para poder
            // hacer una lectura comoda (disponer del metodo readLine()).
            //String fichero = getClass().getResource("/resources/base_datos_sectarios.txt").getPath(); // < para Linux desde el IDE
            //String fichero = "C:\\Users\\pc\\Desktop\\Napakalaki\\src\\resources\\base_datos_sectarios.txt"; // < para Windows
            String fichero = "/resources/base_datos_sectarios.txt"; // < para Linux desde fichero JAR
            File file = this.ficheroExterno(fichero); // Creamos el fichero
            fr = new FileReader(file); // Creamos el manejador de ficheros
            br = new BufferedReader(fr); // Creamos el buffer de lectura para el fichero
            br.readLine(); // lo usamos para leer las cabeceras de las columnas 
            while ((fila = br.readLine()) != null) {
                columnas = fila.split(",");
                mUnusedCultists.add(new Cultist(columnas[0], Integer.parseInt(columnas[1])));
            }
            this.shuffleCultists();
            fr.close();
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
    }

    /**
     * Con este metodo lo que hacemos es convertir una cadena de texto separada
     * por "," en un ArrayList<TreasureKind> de esa manera podemos tener los
     * objetos que realmente necesitamos en el ArrayList devuelto.
     *
     * @param datos una cadena de texto separada por comas con los objetos
     * @return Es un ArrayList<TreasureKind> que contiene los objetos
     */
    private ArrayList<TreasureKind> leeTesoros(String datos) {
        String[] objTesoro = datos.toUpperCase().split("-");
        ArrayList<TreasureKind> Tesoros = new ArrayList<>();

        for (String miTesoro : objTesoro) {
            if (!miTesoro.isEmpty()) {
                Tesoros.add(TreasureKind.valueOf(miTesoro));
            }
        }
        return Tesoros;
    }

    /**
     * Con este metodo barajamos las cartas del monton de los Tesoros.
     */
    private void shuffleTreasures() {
        Collections.shuffle(this.mUnusedTreasures);
    }

    /**
     * Con este metodo barajamos las cartas del monton de los monstruos.
     */
    private void shuffleMonsters() {
        Collections.shuffle(this.mUnusedMonsters);
    }

    /**
     * Con este metodo barajamos las cartas del monton de los Sectarios.
     */
    private void shuffleCultists() {
        Collections.shuffle(this.mUnusedCultists);
    }

    /**
     * Metodo que sirve para poder obtener la unica instancia del manejador de
     * cartas que tendrá el juego.
     *
     * @return la instancia del CardDealer
     */
    public static CardDealer getInstance() {
        return INSTANCE;
    }

    /**
     * Con este metodo conseguiremos que nos devuelvan el siguiente tesoro que
     * no se este usando.
     *
     * @return el siguiente Tesoro
     */
    public Treasure nextTreasure() {
        if (this.mUnusedTreasures.isEmpty()) {
            this.mUnusedTreasures = new ArrayList<>(this.mUsedTreasures);
            this.shuffleTreasures();
            this.mUsedTreasures.clear();
        }
        Treasure tesoro = this.mUnusedTreasures.get(0);
        this.mUnusedTreasures.remove(0);
        return tesoro;
    }

    /**
     * Con este metodo conseguiremos que nos devuelvan el siguiente monstruo que
     * no se este usando.
     *
     * @return el siguiente Monstruo
     */
    public Monster nextMonster() {
        if (this.mUnusedMonsters.isEmpty()) {
            this.mUnusedMonsters = new ArrayList<>(this.mUsedMonsters);
            this.shuffleMonsters();
            this.mUsedMonsters.clear();
        }
        Monster monstruo = this.mUnusedMonsters.get(0);
        this.mUnusedMonsters.remove(0);
        return monstruo;
    }

    /**
     * Con este metodo conseguiremos que nos devuelvan el siguiente player
     * Sectario que no se este usando.
     *
     * @return el siguiente Carta del Sectario
     */
    public Cultist nextCultists() {
        if (this.mUnusedCultists.isEmpty()) {
            this.mUnusedCultists = new ArrayList<>(this.mUsedCultists);
            this.shuffleCultists();
            this.mUsedCultists.clear();
        }
        Cultist cultist = this.mUnusedCultists.get(0);
        this.mUnusedCultists.remove(0);
        return cultist;
    }

    /**
     * Con este metodo conseguiremos insertar en los tesoros usuados el tesoro
     * cuyo parámetro sea pT
     *
     * @param t de tipo Treasure es el tesoro que vamos a usar
     */
    public void giveTreasureBack(Treasure t) {
        this.mUsedTreasures.add(t);
    }

    /**
     * Con este metodo conseguiremos insertar en los tesoros usuados el tesoro
     * cuyo parámetro sea pT
     *
     * @param m de tipo Monstruo es el monstruo que vamos a usar
     */
    public void giveMonsterBack(Monster m) {
        this.mUsedMonsters.add(m);
    }

    /**
     * Sirve para poder construir la bases de datos de cartas y barajarlas para
     * el juego
     */
    public void initCards() {
        this.initMonsterCardDeck();
        this.initTreasureCardDeck();
        this.initCultistCardDeck();
    }

    /**
     * Devuelve el array completo de monstruos que estamos usando.
     *
     * @return el parametro mUsedMonsters
     */
    public ArrayList<Monster> getUsedMonsters() {
        return this.mUsedMonsters;
    }

    /**
     * Devuelve el array completo de los Treasures que estamos usuando.
     *
     * @return el parametro mUsedTreasures
     */
    public ArrayList<Treasure> getUsedTreasures() {
        return this.mUsedTreasures;
    }
}
